### PR TITLE
Validate limits

### PR DESCRIPTION
**Pull request by alexsnaps**
This introduces the option `--validate` to the `limitador-server` executable.

It does validate the `LIMITS_FILE` passed as argument against the expected syntax. 
Prints any error to `stderr`
Prints the validated output back to `stdout`

<details>
## When there is an error:

```
 ./target/debug/limitador-server --validate limitador-server/examples/limits.yaml > new.yaml
Invalid file: ConfigFile("Couldn't parse: .[0]: missing field `seconds` at line 3 column 3")
```

## When there is a condition error:

```
./target/debug/limitador-server --validate limitador-server/examples/limits.yaml > new.yaml
Invalid file: ConfigFile("Couldn't parse: .[0].conditions: SyntaxError: Invalid character `~` at offset 12 of condition \"req.method ~= GET\" at line 7 column 5")
```

## When used to port deprecated syntax to the new:

```
./target/debug/limitador-server --validate limitador-server/examples/limits.yaml > new.yaml
Deprecated syntax for conditions corrected!
```

with ` limitador-server/examples/limits.yaml`

```yaml
---
-
  namespace: test_namespace
  max_value: 10
  seconds: 60
  conditions:
    - "req.method == GET"
  variables:
    - user_id
-
  namespace: test_namespace
  max_value: 5
  seconds: 60
  conditions:
    - "req.method == POST"
  variables:
    - user_id
```

and the resulting ported `new.yaml`

```yaml
- namespace: test_namespace
  max_value: 10
  seconds: 60
  name: null
  conditions:
  - req.method == "GET"
  variables:
  - user_id
- namespace: test_namespace
  max_value: 5
  seconds: 60
  name: null
  conditions:
  - req.method == "POST"
  variables:
  - user_id
```

</details>

### CODE REVIEW

**Comment by alexsnaps** [comment]
See #122 , but should I add the check of `limit.max_value > 0` as an addition check as well as part of this?

**Review by eguzki — CHANGES_REQUESTED** [review]
LGTM

File: limitador-server/examples/limits.yaml [hunk 1]
```diff
@@ -12,6 +12,6 @@
   max_value: 5
   seconds: 60
   conditions:
-    - "req.method == 'POST'"
+    - "req.method ~= 'POST'"
```
[the remark this answers is not in the document]
Comment (alexsnaps) [reply limitador-server/examples/limits.yaml:15]: Ooops… Yeah, that was not meant to be committed 🤦

**Comment by alexsnaps** [comment]
@eguzki fixed your catch… and added 894fcba - let me know what you think… This will not allow zero (`0`) and negative values for `Limit.max_value`… Tho I still don't know why this isn't a `u64` to begin with… 🤔 Or did I miss/forget something?

**Comment by alexsnaps** [comment]
@eguzki and I think this addresses all we've discussed in the status meeting

File: limitador-server/src/main.rs [hunk 2]
```diff
@@ -205,16 +205,22 @@ impl Limiter {
...
                 let parsed_limits: Result<Vec<Limit>, _> = serde_yaml::from_reader(f);
                 match parsed_limits {
-                    Ok(limits) => {
-                        match &self {
-                            Self::Blocking(limiter) => limiter.configure_with(limits)?,
-                            Self::Async(limiter) => limiter.configure_with(limits).await?,
-                        }
-                        if limitador::limit::check_deprecated_syntax_usages_and_reset() {
-                            error!("You are using deprecated syntax for your conditions! See the migration guide https://kudrant.io/migration/limitador/constraints")
+                    Ok(limits) => match first_negative(&limits) {
```
Comment (eguzki) [inline limitador-server/src/main.rs:208]: missing some test with negative values?

File: limitador-server/src/main.rs [hunk 3]
```diff
@@ -462,6 +477,12 @@ fn create_config() -> (Configuration, String) {
                 .display_order(6)
                 .help("Sets the level of verbosity"),
         )
+        .arg(
+            Arg::with_name("validate")
```
Comment (eguzki) [inline limitador-server/src/main.rs:481]: This will be documented as part of the limitador-server v1.0.0 crate, right?

**Review by eguzki — CHANGES_REQUESTED** [review]

File: limitador-server/src/main.rs (hunk 3 above)
Comment (alexsnaps) [reply limitador-server/src/main.rs:481]: There is no such crate. But yes, there is task on-going to document _all_ these things. This will also be in the migration guide.

File: limitador-server/src/main.rs (hunk 2 above)
Comment (alexsnaps) [reply limitador-server/src/main.rs:208]: I can add a test for that function being called if that's what you mean

File: limitador-server/src/main.rs [hunk 4]
```diff
@@ -552,6 +573,41 @@ fn create_config() -> (Configuration, String) {
...
 
     let limits_file = matches.value_of("LIMITS_FILE").unwrap();
+
+    if matches.is_present("validate") {
+        let error = match std::fs::File::open(limits_file) {
+            Ok(f) => {
+                let parsed_limits: Result<Vec<Limit>, _> = serde_yaml::from_reader(f);
+                match parsed_limits {
+                    Ok(limits) => {
+                        if limitador::limit::check_deprecated_syntax_usages_and_reset() {
```
Comment (eguzki) [inline limitador-server/src/main.rs:583]: should `first_negative` be called to check existence of negative maxValues?

Comment (alexsnaps) [reply limitador-server/src/main.rs:583]: Well it does return the first negative `max_value` as an `Option<usize>`

Comment (alexsnaps) [reply limitador-server/src/main.rs:583]: `find_first_negative_limit` maybe?

Comment (eguzki) [reply limitador-server/src/main.rs:583]: my point is, the validation command should also check there is no "negative limits", right?

Comment (alexsnaps) [reply limitador-server/src/main.rs:583]: Oh! Yes!!!

**Review by eguzki — APPROVED** [review]
